File: e2e/factories/optionPools.ts
import { db } from "@test/db";
import { companiesFactory } from "@test/factories/companies";
import { shareClassesFactory } from "@test/factories/shareClasses";
import { optionPools } from "@/db/schema";
import { assert } from "@/utils/assert";

export const optionPoolsFactory = {
  create: async (overrides: Partial<typeof optionPools.$inferInsert> = {}) => {
    const [optionPool] = await db
      .insert(optionPools)
      .values({
        companyId: overrides.companyId || (await companiesFactory.create()).company.id,
        shareClassId: overrides.shareClassId || (await shareClassesFactory.create()).shareClass.id,
        name: overrides.name || "Best equity plan",
        authorizedShares: overrides.authorizedShares || 100n,
        issuedShares: overrides.issuedShares || 50n,
        ...overrides,
      })
      .returning();
    assert(optionPool != null);

    return { optionPool };
  },
};

File: backend/spec/factories/option_pools.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :option_pool do
    company
    share_class
    name { "Best equity plan" }
    authorized_shares { 100 }
    issued_shares { 50 }

    # So that `available_shares` is loaded correctly
    after(:create) {  _1.reload }
  end
end
